packages/server-wallet/src/wallet/index.ts

```diff
@@ -617,8 +619,18 @@ export class Wallet extends EventEmitter<WalletEvent>
           this.walletConfig.timingMetrics
         );
 
-        if (!nextAction) markObjectiveAsDone();
-        else {
+        if (!nextAction) {
+          markObjectiveAsDone();
+          // todo: is there a better check for "has the final postfund state been signed"?
+          if (app.supported?.turnNum === app.participants.length * 2 - 1) {
+            const toEmit: SingleChannelOutput = {
+              channelResult: ChannelState.toChannelResult(app),
+              outbox: [],
+            };
+            const channelReady: WalletEventName = 'channelReady';
+            this.emit(channelReady, toEmit);
+          }
```

Comment:
In the description of the PR you characterise the problem as a promise resolving before the channel is open, and therefore the need to separately notify the app when the channel is ready. It seems to me that the problem goes beyond this.

I'd characterise the problem as "we need to tell the CM that it needs to act when the next step of the protocol is triggered by a chain-service event". When the deposit event that fully funds the channel arrives, we need to (a) trigger a post-fund-setup to be sent to the opponent, and (b) when all messages have been exchanged, notify the app that funding is complete. It seems that to accomplish (a), you need to have a notification that also includes messages to send. The interaction would be like this:

1. ChainService sends funding event to wallet
2. Wallet updates funding on channels and runs `takeActions`
3. Take actions realises that the next step is to send the post-fund-0
4. Wallet sends an event to the CM, containing both a channel-result and the post-fund-0 to send to the opponent
    * ☝️  **this is the bit that we're currently missing (I think?)**
5. CM sends to RM, creating a promise that resolves when all messaging is done
6. When the promise resolves, then the channel will be open, and the CM will be given the channel result

It seems that if you solve (a), then you don't need the notification here - you need notifications to be emitted exactly when activity is triggered by a chain service event. If that's true, it seems worth removing this event emission, as one of the confusions in the xstate wallet is when the double notifications that the app would receive about some events - one through a notification, and the other from the return value.